### PR TITLE
Redundant thread compute analysis to avoid un-necessary sync insertion

### PR DESCRIPTION
**Pull request by shmsong**
This is a follow up on #1684. An analysis traversal pass is added to make sure sync on thread redundant writes are inserted only when needed.

### CODE REVIEW

File: torch/csrc/jit/codegen/cuda/lower_sync_information.cpp [hunk 1]
```diff
@@ -205,25 +211,25 @@ void SyncMap::build(Fusion* fusion) {
...
           }
 
+          // In the case when the parallel id's are mapped by ca map,
+          //   will additionally need to consider if the producer is
+          //   a redundant write. The raw dim can be skipped only if
+          //   consumer use chains only contain redundant uses.
+          //  TODO:
+          //    still losing a bit precision here for expr ordering
+          //  sensitive cases, but we could wait until that becomes
+          //  a perf limiter to fix.
```
Comment (shmsong) [inline torch/csrc/jit/codegen/cuda/lower_sync_information.cpp:225]: This analysis is expression order insensitive so we'd need to incorporate this TODO into the sync insertion pass, which didn't seem to be trivial to do, essentially need to track the smem writes more carefully. 

As this case seems to be a rare case among already rare cases, in terms of perf benefit I think it could wait.

File: torch/csrc/jit/codegen/cuda/fusion.h [hunk 2]
```diff
@@ -153,6 +153,14 @@ class TORCH_CUDA_CU_API Fusion : public IrContainer {
   //! also included as they must show up in the final code.
   std::vector<Val*> usedMathVals();
 
+  //! Returns all vals that are produced by used math expressions and
+  //!  also do not have further consumers.
+  //!
+  //! In the case of an active multi-output expressions, the returned vector
+  //!  will include the expression outputs that did not lead to an fusion
+  //!  output.
+  std::vector<Val*> terminatingMathVals();
```
Comment (naoyam) [inline torch/csrc/jit/codegen/cuda/fusion.h:162]: What's the difference with `getTerminatingOutputs`? Examples would be helpful.

**Comment by naoyam** [comment]
Can you show a simple example fusion with the previous and new generated code? Would be helpful to remind me what the issue was.

**Comment by shmsong** [comment]
> Can you show a simple example fusion with the previous and new generated code? Would be helpful to remind me what the issue was.

This PR basically removes the redundant sync mentioned in here:
[https://github.com/csarofeen/pytorch/pull/1684#issuecomment-1121687568](https://github.com/csarofeen/pytorch/pull/1684#issuecomment-1121687568)

File: torch/csrc/jit/codegen/cuda/fusion.h (hunk 2 above)
Comment (shmsong) [reply torch/csrc/jit/codegen/cuda/fusion.h:162]: It's just getting all the dangling vals that does not have any uses. Simple case: 

```
var, avg, n = welford(in)

addOutput(var)
```

terminating output -> `var`
terminating vals -> `var, avg, n`

This is useful when we want to do a backward visitor pass.

Comment (shmsong) [reply torch/csrc/jit/codegen/cuda/fusion.h:162]: The feature this PR enables is lower priority but it's kind of interesting that this is the first backward visitor pass on TV ops as far as I can tell. So some infrastructure needed to be set up.

File: torch/csrc/jit/codegen/cuda/lower_thread_predicate.cpp [hunk 3]
```diff
@@ -285,6 +285,154 @@ void ThreadPredicateMap::updateBitSet(const Expr* expr) {
   }
 }
 
+namespace {
+
+//! A simple backward data flow pass:
+//!  This pass propagates information backward to annotate "redundant use
+//!  chain"'s.
+//! The reason this is needed is that if we have a chain of register-to-register
```
Comment (naoyam) [inline torch/csrc/jit/codegen/cuda/lower_thread_predicate.cpp]: Is this propagation analysis strictly for register-to-register expr chains? If so, why there's a propagation rule for shared memory tensors?

File: torch/csrc/jit/codegen/cuda/lower_thread_predicate.cpp [hunk 4]
```diff
@@ -285,6 +285,154 @@ void ThreadPredicateMap::updateBitSet(const Expr* expr) {
...
+      return c10::nullopt;
+    } else {
+      return redundant_map_it->second;
+    }
+  }
+
+ private:
+  using BackwardVisitor::handle;
+
+  void handle(TensorView* tv) final {
```
Comment (naoyam) [inline torch/csrc/jit/codegen/cuda/lower_thread_predicate.cpp:350]: If I read this function correctly, it seems that the `redundant_types` of `tv`, when it's shared or local, doesn't matter and `redundant_tv_use_map_[tv]` is solely determined by the consumers of `tv`. Is it correct?

File: torch/csrc/jit/codegen/cuda/lower_thread_predicate.cpp (hunk 3 above)
Comment (shmsong) [reply torch/csrc/jit/codegen/cuda/lower_thread_predicate.cpp]: Shared memory part is trying to include analysis for BID as well so this info will be there when we need it, since we could in the same nature have redundant block writes as well for intermediate gmem tensors, though I don't think we have been exercising global sync insertion too much at the moment.

File: torch/csrc/jit/codegen/cuda/lower_thread_predicate.cpp (hunk 4 above)
Comment (shmsong) [reply torch/csrc/jit/codegen/cuda/lower_thread_predicate.cpp:350]: I think that's correct. 

`redundant_types` is the same definition as before.

`redundant_tv_use_map_` is the new information that is derived from the `redundant_types` of the consumer chains.

Comment (naoyam) [reply torch/csrc/jit/codegen/cuda/lower_thread_predicate.cpp:350]: So, a TV can have `redundant_tv_use` parallel types even if it doesn't have anything in `redundant_types`?

File: torch/csrc/jit/codegen/cuda/lower_thread_predicate.cpp (hunk 3 above)
Comment (naoyam) [reply torch/csrc/jit/codegen/cuda/lower_thread_predicate.cpp]: OK, but the comment seems confusing then?

Comment (shmsong) [reply torch/csrc/jit/codegen/cuda/lower_thread_predicate.cpp]: Ah, yes it seems outdated. Just updated it. Thanks!

File: torch/csrc/jit/codegen/cuda/lower_thread_predicate.cpp (hunk 4 above)
Comment (shmsong) [reply torch/csrc/jit/codegen/cuda/lower_thread_predicate.cpp:350]: Yes. The `redundant write` and `redundant read` are two pieces of info that complement each other and each could be useful in different scenarios.

In the context of this PR, we can remove the sync when all `redundant write` are also `redundant read`

Comment (naoyam) [reply torch/csrc/jit/codegen/cuda/lower_thread_predicate.cpp:350]: Hmm, I'm still confused. In the header file, it says:

```
Tracking use chain of redundant writes:
```

If `a TV can have redundant_tv_use parallel types even if it doesn't have anything in redundant_types` is true, it seems it tracks use chains of non-redundant writes as well. What am I missing?

Comment (shmsong) [reply torch/csrc/jit/codegen/cuda/lower_thread_predicate.cpp:350]: Oh that looks like confusing wording, would change it to `use chains that end with redundant writes`. 

The information it conveys is different from `redundant writes` and it is totally valid for a tv that isn't redundantly written, i.e. all the threads participate in writing the tv into shared mem but say only threads with `tidx==0` actually use it. Scenario like this is modeled in the analysis though it's not really used in this PR.

Comment (naoyam) [reply torch/csrc/jit/codegen/cuda/lower_thread_predicate.cpp:350]: OK, I guess I'm slowly getting this, but calling it "redundant reads" still sounds confusing. For example, if a tensor is never used but is just output, and it is redundantly written, the algorithm would make it as also redundantly read, even though it's never read, right?

Could you add some unit tests just for this pass? Some small tests demonstrating its intentions as well as exercising corner cases would be really helpful to get the gist of this analysis.

Comment (shmsong) [reply torch/csrc/jit/codegen/cuda/lower_thread_predicate.cpp:350]: Was reviewing the code again, and now I think the major confusion comes from the way the traversal is viewed. The `redundant use` info in the predicateInfo is conceptually one iteration earlier than `redundant type` originally. So `redundant consumer` actually meant `redundant as a consumer`. So in your example is an output is redundantly written, then it's producer can skip syncs if the output is the only consumer since the output is `redundant as a consumer`. 

I guess the reason it was done this way originally was to use it as a drop in fix in sync info pass, which actually does look into immediate consumers, where this info was applied.

Now I agree it's more confusing this way and I will just add a final iteration in this building pass to aggregate for each tv so that the `redundant use` and `redundant write` look conceptually more symmetric, and in this case an output would not be "redundantly used" : ).

Also definitely I could add more unit tests on the info pass.

**Comment by csarofeen** [comment]
@shmsong @naoyam is this PR still alive? What's it waiting on?

**Comment by naoyam** [comment]
> @shmsong @naoyam is this PR still alive? What's it waiting on?

Last I checked, @shmsong was saying he would do some refactoring. Not sure if it's done. @shmsong, is it ready to review again?

**Comment by shmsong** [comment]
> > @shmsong @naoyam is this PR still alive? What's it waiting on?
> 
> Last I checked, @shmsong was saying he would do some refactoring. Not sure if it's done. @shmsong, is it ready to review again?

Yes just checked. I think I pushed all the changes I wanted to make. It should be ready for review again.

**Comment by csarofeen** [comment]
@naoyam please review.

File: torch/csrc/jit/codegen/cuda/lower_thread_predicate.h [hunk 5]
```diff
@@ -48,9 +48,24 @@ class TORCH_CUDA_CU_API ThreadPredicateMap {
...
+    //  [Redundant use chain]
+    //  a parallel type is a `redundant_consumer_type` only
+    //    if all of its propagation use chains terminate with
+    //    a redundant write of this type.
+    //  A propagation use chain is currently either a reg-to-reg
+    //   chain for a shared mem tv, or a reg/smem-to-reg/smem chain
+    //   for a global tv.
+    // This is complementary information to `redundant_types`.
+    //  If a tensor view is redundantly written and not fully
+    //  redundantly used, see FusionRedundantPredSync3,
```
Comment (naoyam) [inline torch/csrc/jit/codegen/cuda/lower_thread_predicate.h]: I couldn't immediately get what "not fully redundantly used" means. Maybe just something like "unless redundantly used by all consumers".

File: torch/csrc/jit/codegen/cuda/lower_thread_predicate.cpp [hunk 6]
```diff
@@ -285,6 +285,200 @@ void ThreadPredicateMap::updateBitSet(const Expr* expr) {
...
+  std::unordered_map<const TensorView*, ParallelTypeBitmap>
+      redundant_consumer_parallel_type_map;
+
+  // This map is complementary information to redundant type.
+  //  It indicates if a tv is only redundantly consumed in any
+  //   parallel dimension. An example would be T0_shared in the
+  //  example above, which would is only redundantly consumed
+  //   in tidx.
+  std::unordered_map<const TensorView*, ParallelTypeBitmap>
+      redundant_tv_use_map;
```
Comment (naoyam) [inline torch/csrc/jit/codegen/cuda/lower_thread_predicate.cpp]: nit: `redundant_tv_use_map_`

File: torch/csrc/jit/codegen/cuda/lower_thread_predicate.cpp [hunk 7]
```diff
@@ -285,6 +285,200 @@ void ThreadPredicateMap::updateBitSet(const Expr* expr) {
...
+  //  T1_local = T0_shared[...]
+  //  if(tid.x == 0)
+  //    T2_shared[...] = T1_local[...]
+  // Then tidx would be redundant consumer parallel type
+  //  for T1, as T1 is local tensor, and only threads satisfying
+  //  tidx == 0 would need to provide a valid data.
+  // In this case, not all threads would need to read correct data
+  //  from T0_shared, which would help remove some sync's.
+  std::unordered_map<const TensorView*, ParallelTypeBitmap>
+      redundant_consumer_parallel_type_map;
```
Comment (naoyam) [inline torch/csrc/jit/codegen/cuda/lower_thread_predicate.cpp]: nit: `redundant_consumer_parallel_type_map_`

File: torch/csrc/jit/codegen/cuda/lower_thread_predicate.h (hunk 5 above)
Comment (shmsong) [reply torch/csrc/jit/codegen/cuda/lower_thread_predicate.h]: Updated. Thanks!

File: torch/csrc/jit/codegen/cuda/lower_thread_predicate.cpp [hunk 8]
```diff
@@ -285,6 +285,200 @@ void ThreadPredicateMap::updateBitSet(const Expr* expr) {
...
+//!          The parallel type is redundantly used in the expr input
+//!          only if all of the outputs redundantly use the same type.
+class RedundantUseAnalysis : BackwardVisitor {
+ public:
+  RedundantUseAnalysis(Fusion* fusion, const ThreadPredicateMap& pred_map)
+      : fusion_(fusion), pred_map_(pred_map) {
+    traverseFrom(fusion, fusion->terminatingMathVals());
+  }
+
+  c10::optional<ParallelTypeBitmap> getRedundantUseBitMap(
```
Comment (naoyam) [inline torch/csrc/jit/codegen/cuda/lower_thread_predicate.cpp]: Can this just return an empty `ParallelTypeBitmap`?

File: torch/csrc/jit/codegen/cuda/lower_thread_predicate.cpp [hunk 9]
```diff
@@ -285,6 +285,200 @@ void ThreadPredicateMap::updateBitSet(const Expr* expr) {
...
+        }
+      }
+
+      TORCH_INTERNAL_ASSERT(
+          maybe_expr_pred_map.has_value(), "TV op not having a tv output");
+      redundant_expr_use_map_[expr] = maybe_expr_pred_map.value();
+    }
+  }
+
+ private:
```
Comment (naoyam) [inline torch/csrc/jit/codegen/cuda/lower_thread_predicate.cpp:438]: I find it difficult to remember what the difference is between `redundant_consumer_parallel_type_map` and `redundant_tv_use_map`. I'm sorry I don't have any specific suggestion, but it seems to me these two names give very little info to help differentiate the two maps.

**Review by naoyam — APPROVED** [review]
LGTM. Thanks for addressing the issue.

Left a couple of minor comments, mostly just stylistic concerns. E.g., is `c10::optional` really necessary? Also, would be great if the two maps of `RedundantUseAnalysis` were cleaned up somehow so that it would become clear how they were used.  Maybe, we could just remove `redundant_use_map` and dynamically compute them inside `getRedundantUseBitMap`?

File: torch/csrc/jit/codegen/cuda/lower_thread_predicate.cpp (hunk 6 above)
Comment (shmsong) [reply torch/csrc/jit/codegen/cuda/lower_thread_predicate.cpp]: Updated. Thanks!

File: torch/csrc/jit/codegen/cuda/lower_thread_predicate.cpp (hunk 7 above)
Comment (shmsong) [reply torch/csrc/jit/codegen/cuda/lower_thread_predicate.cpp]: Updated. Thanks!

File: torch/csrc/jit/codegen/cuda/lower_thread_predicate.cpp (hunk 8 above)
Comment (shmsong) [reply torch/csrc/jit/codegen/cuda/lower_thread_predicate.cpp]: Yes. Given the only place this is used. This would simplify the logic. Thanks!

File: torch/csrc/jit/codegen/cuda/lower_thread_predicate.cpp (hunk 9 above)
Comment (shmsong) [reply torch/csrc/jit/codegen/cuda/lower_thread_predicate.cpp:438]: Yes I agree this is confusing. I just removed `redundant_tv_use_map` and compute it where it is used.
